### ISSUE TITLE
Haskell Syntax highlight break when write a type signature with "=>" sign

### ISSUE THREAD

**Issue by foxmean** [helix-editor/helix#11517]
### Summary

Around 95% up when I write a type signature with "=>" sign, the syntax highlight is break. Everything is recover when I quite and reopen the helix as show below.
[![asciicast](https://asciinema.org/a/1TB081PmRJ3BTCzfmQvxbesSX.svg)](https://asciinema.org/a/1TB081PmRJ3BTCzfmQvxbesSX)

### Reproduction Steps

 

I tried this:

1. Open the helix editor and write some Haskell script
2. Everything worked just fine except when I wrote a type signature that included "=>" sign (mostly). I usually not happen when I haven't included "=>".
3. When I save and restart helix, everything fixed. But it is annoying.

I expected this to happen:
The syntax highlighter maybe break due to incomplete writing since it haven't happened if I wrote the function definition *before* type signature as show below
[![asciicast](https://asciinema.org/a/nNTGRRhXe4vY2wGrl0FKzkADv.svg)](https://asciinema.org/a/nNTGRRhXe4vY2wGrl0FKzkADv) 

Instead, this happened:
When I wrote a type signature before the  function definition
[![asciicast](https://asciinema.org/a/1TB081PmRJ3BTCzfmQvxbesSX.svg)](https://asciinema.org/a/1TB081PmRJ3BTCzfmQvxbesSX)

### Helix log

A copy of `~/.cache/helix/helix.log` here (with `RUST_BACKTRAACE=1 hx -vv` when reproducing bug. I have no idea how to summary these lines but I have selected only line that I reproduced the bug [here](https://pastecode.io/s/rxou198h). 

### Platform

MacOS Sonoma 14.6.1; on MacBook Air M1, 2020

### Terminal Emulator

iTerm2, Build 3.5.0 (without terminal multiplex); and Alacritty Version 0.13.2 (1) with Zellij 0.40.1. Both terminal was installed via homebrew cask. Zellij was installed via nixpkgs (nix-darwin)

### Installation Method

nixpkgs through nix-darwin flake

### Helix Version

helix 24.7 (079f5442)

**Comment by kirawi** [comment]
We disable highlighting on a buffer if highlighting takes longer than half a second to compute. It sounds to me like `=>` has some buggy behavior here. We could see if a newer commit for the grammar has resolved this issue.

**Comment by foxmean** [comment]
Sorry for the last archived video. I can reproduce some of the bug related to just "->" too like.
[![asciicast](https://asciinema.org/a/1DoAM5Xz6NrvQ0BPLblBmWWry.svg)](https://asciinema.org/a/1DoAM5Xz6NrvQ0BPLblBmWWry)
BTW, I think the bug is from HLS itself due to I can reproduce it in Zed editor.
I use GHC 9.10.1 with HLS 2.9.0.1 from GHCup

**Comment by foxmean** [comment]
> We disable highlighting on a buffer if highlighting takes longer than half a second to compute. It sounds to me like `=>` has some buggy behavior here. We could see if a newer commit for the grammar has resolved this issue.

Where is the place that I should follow for this issue or take a look if someone fix it already in development build or anything I can do for help fixing it?

**Comment by foxmean** [comment]
I've try to change `timeout` in `[language-server.haskell-language-server]` as in [the documentation](https://docs.helix-editor.com/languages.html#language-server-configuration) specified to something like 6_000_000 but the bug still wasn't fixed.

**Comment by RoloEdits** [comment]
You can try to update the tree sitter grammar for Haskell to see if there a parser error was fixed. 

https://github.com/tree-sitter/tree-sitter-haskell

Current rev if from `Sep 24, 2023`.

**Comment by foxmean** [comment]
One thing I found wired is that I cannot reproduce it in Neovim (NVIM v0.10.2, Build type: Release, LuaJIT 2.1.1731601260). But consistently reproducible with Helix: video: [![asciicast](https://asciinema.org/a/6RQ8SIpk59QhAuqVD5oFmjvFz.svg)](https://asciinema.org/a/6RQ8SIpk59QhAuqVD5oFmjvFz) .

**Comment by david-d-h** [comment]
+1 still happening for me, doesn't happen if the type definition is the only thing in the file
haven't tested in other editors

**Comment by foxmean** [comment]
I am currently have no skill to tackle this problem which I think it is about Haskell's tree-sitter grammar rewriting, since this bug is also reproducible with other editor(s) like Zed. However, I've found another workaround (if you don't want to just quit & reopen the file) for Helix: you can use `:config-reload` command and the syntax highlighting will be back.

**Comment by CalebLarsen** [comment]
@foxmean Will you try building from source and let me know if you are still experiencing this issue?
[#12972](https://github.com/helix-editor/helix/pull/12972) was a huge change to how we work with tree-sitter.

**Comment by foxmean** [comment]
> [@foxmean](https://github.com/foxmean) Will you try building from source and let me know if you are still experiencing this issue? [#12972](https://github.com/helix-editor/helix/pull/12972) was a huge change to how we work with tree-sitter.

Yeah, it works! (I've tested `bcb6c20a`. I think it is the same as this one bcb6c20a84b6ac6697bf2a4b5d0b78832bcfea62) Thank you so much for you and helix team. Waiting for next official release. Should I close this issue?

[![asciicast](https://asciinema.org/a/725889.svg)](https://asciinema.org/a/725889)

**Comment by CalebLarsen** [comment]
I think we're good to close the issue, thank you!
